grabbit/src/main/groovy/com/twcable/grabbit/server/batch/steps/jcrnodes/JcrNodesProcessor.groovy

```diff
@@ -88,10 +90,51 @@ class JcrNodesProcessor implements ItemProcessor<JcrNode, Node> {
             }
         }
 
-        final List<JcrProperty> properties = jcrNode.properties.toList()
-        Node.Builder nodeBuilder = Node.newBuilder()
+        // It checks if the node parent has required primary type as "nt:hierarchyNode"
+        // and then skip it as the node has already been processed by parent node
+        if(isRequiredHierarchyChildNode(jcrNode)) {
+            return null;
+        } else if(isHierarchyNode(jcrNode)) {
+            // Parent "nt:hierarchyNode" type will be processed
+            return buildNode(jcrNode, getRequiredChildNodes(jcrNode))
+        } else {
+            // Non-Hierarchy node logic
+            return buildNode(jcrNode, null)
+        }
+    }
+
+    private Node buildNode(JcrNode jcrNode, Collection<JcrNode> childNodes) {
+        final Node.Builder nodeBuilder = Node.newBuilder()
         nodeBuilder.setName(jcrNode.path)
+        nodeBuilder.setProperties(buildProperties(jcrNode))
+        childNodes?.each {
+            nodeBuilder.addChildNode(buildNode(it, getRequiredChildNodes(it)))
+        }
+        return nodeBuilder.build()
+    }
+
+    private static Collection<JcrNode> getRequiredChildNodes(JcrNode jcrNode) {
+        // Here we chose not to handle the edge case of a required child hierarchy node requiring a grandchild node
+        // because no node structure exists like that within AEM.
+        return jcrNode.getNodes()
+                .findAll{JcrNode childJcrNode -> isRequiredHierarchyChildNode(childJcrNode)}
+    }
 
+    private static boolean isRequiredHierarchyChildNode(JcrNode node) {
```

Comment:
Just spoke with Jim around some findings.  It doesn't look like it's really nt:hierarchyNode types we are looking for.  If you go to your CQ's /crx/de/exportnodetype.jsp ; and navigate to nt:hierarchyNode, you'll see that it is essentially an empty semantic abstract type.  You can also find other unstructured nodes that have required mandatory children, such as rep:system (as an example).  

Rather than looking for nt:hierarchyNodes, we may need to see if a node has mandatory children/names via http://www.day.com/maven/jsr170/javadocs/jcr-2.0/javax/jcr/nodetype/NodeDefinition.html#getRequiredPrimaryTypes() and http://www.day.com/maven/jsr170/javadocs/jcr-2.0/javax/jcr/nodetype/NodeDefinition.html#getRequiredPrimaryTypeNames()